rclcpp/test/test_timer.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <exception>
+#include <memory>
+
+#include "rcl/timer.h"
+
+#include "rclcpp/rclcpp.hpp"
+
+using namespace std::chrono_literals;
+
+class TestTimer : public ::testing::Test
+{
+protected:
+  void SetUp()
+  {
+    rclcpp::init(0, nullptr);
+    test_node = std::make_shared<rclcpp::Node>("test_timer_node");
+  }
+
+  void TearDown()
+  {
+    test_node.reset();
+    rclcpp::shutdown();
+  }
+
+  rclcpp::Node::SharedPtr test_node;
+};
+
+TEST_F(TestTimer, test_is_cancelled)
```

Comment:
Having a single test case is a code smell:
https://testing.googleblog.com/2018/06/testing-on-toilet-keep-tests-focused.html

If you end up having:

```c++
TEST(...) {
   // I do something
   {
      // some asserts
   }
   // I do something else
   {
      // more asserts
   }
}
```

You basically wrote two tests cases in one and those two should be separated.

I.e. your reset and cancel should be a different test case IMHO

Reply:
I don't agree, as the current test is valid*. Adding a more permutations appears to be appropriate. 

*Checking that the timer state was reset after a cancellation, checking timer state after cancellation, and checking after a reset are atomic, valid permutations of tests.

Reply:
Ok, that makes sense.